File: Numerical Computing Library/Matrix/src/numerical/Eigen.cpp
#include "../../include/numerical/Eigen.hpp"
#include <iostream>
#include <cmath>

void Eigen::gershgorin(const Matrix& A) {
    int n = A.getRows();

    std::cout << "\nGershgorin Circles:\n";

    for(int i = 0; i < n; i++) {
        double center = A.get(i,i);
        double radius = 0;

        for(int j = 0; j < n; j++) {
            if(i != j)
                radius += std::abs(A.get(i,j));
        }

        std::cout << "Circle " << i+1 << ": Center = "
                  << center << ", Radius = " << radius << "\n";

        std::cout << "Range: [" << center - radius
                  << ", " << center + radius << "]\n\n";
    }
}

File: Numerical Computing Library/NumericalLibrary-ver2/library/MatrixV1_Python/matrixv1/numerical/eigen.py
"""Gershgorin circle theorem — eigenvalue estimates."""

from matrixv1.core.matrix import Matrix


def gershgorin(A: Matrix) -> None:
    """Print Gershgorin disc estimates for eigenvalues of A."""
    if not A.is_square():
        raise ValueError("Gershgorin: matrix must be square")
    n = A.rows
    print("  Gershgorin eigenvalue estimates:")
    for i in range(n):
        centre = A.get(i, i)
        radius = sum(abs(A.get(i, j)) for j in range(n) if j != i)
        print(f"    Row {i}: centre = {centre:.6g},  radius = {radius:.6g}"
              f"  →  λ ∈ [{centre - radius:.6g}, {centre + radius:.6g}]")
